odoo/models.py

```diff
@@ -4113,6 +4113,11 @@ def _write(self, vals):
             field = self._fields[name]
             assert field.store
             assert field.column_type
+
+            if field.type == 'integer' and (val < -2147483648 or val > 2147483647):
+                flow = 'under' if val < 0 else 'over'
+                raise ValueError("Integer field %sflow for column '%s' on model '%s'" % (flow, name, self._name))
```

Comment:
Maybe we can defined this check as a  field value check instead of a SQL value check.
I suggest we do it in the `convert_to_cache` and `convert_to_column` in the fields.py.  I think it would be eaiser to extend a bigint field if necessary in the future.
Note: your current check is only for stored field, which means non-stored fields don't have to repect this check, remember to add a condition in those `convert`

for business, since you want a clear message. Maybe it would be better
1. make the erorr message translatable `_()`
2. show the exact number here? If the number is directly provided by the user, he can know what's wrong.